### PR TITLE
[DYOD] Add null value specification to SyntheticTableGenerator

### PR DESCRIPTION
**Pull request by birneamstiel**
This will add a way to specify which part (ratio) of the generated table will be filled with null values. Since this would add another specification on per-column-basis to `generateTable` we decided to factor out all specifications that apply per column into one vector of a newly introduced struct `ColumnSpecification`.

### TODO
- [x] check for other usages
- [x] add FullCI label

### CODE REVIEW

File: src/benchmarklib/synthetic_table_generator.cpp [hunk 1]
```diff
@@ -149,6 +145,23 @@ std::shared_ptr<Table> SyntheticTableGenerator::generate_table(
...
+          std::vector<bool> null_values;
+
+          /**
+           * If a ratio of to-be-created NULL values is given, fill the null_values vector used in the ValueSegment
+           * constructor in a regular interval based on the null_ratio with true.
+           */
+          if (column_specifications[column_index].null_ratio) {
+            null_values = std::vector<bool>(chunk_size, false);
+
+            const double step_size = 1.0 / column_specifications[column_index].null_ratio.value();
```
Comment (birneamstiel) [inline src/benchmarklib/synthetic_table_generator.cpp]: We should probably use step_size as int here. Check that this still gives the right ratio in the end.

**Comment by tobodner** [comment]
@mrks: Can you please have a quick look at the CI worker for Mac? The ClangDebugMac job failed (with "the remote end hung up unexpectedly").

**Comment by mrks** [comment]
Ah, wait. That's a different thing. That's just the HPI network being the HPI network.

**Comment by jonasnoki** [comment]
@mrks so a simple restart should do the trick?

**Comment by mrks** [comment]
https://hyrise-ci.epic-hpi.de/blue/organizations/jenkins/hyrise%2Fhyrise/detail/PR-1987/8/pipeline

but CI is busy right now

File: src/test/synthetic_table_generator_test.cpp [hunk 2]
```diff
@@ -71,13 +71,10 @@ TEST_P(SyntheticTableGeneratorDataTypeTests, IntegerTable) {
   std::transform(all_segment_encoding_specs.begin(), all_segment_encoding_specs.end(),
                  std::back_inserter(supported_segment_encodings), replace_unsupporting_encoding_types);
 
-  const auto test_data_types = std::vector<DataType>(supported_segment_encodings.size(), tested_data_type);
-  const auto test_data_distributions =
-      std::vector<ColumnDataDistribution>(supported_segment_encodings.size(), std::get<1>(GetParam()));
-  const auto column_names = std::vector<std::string>(supported_segment_encodings.size(), "column_name");
+  ColumnSpecification column_specification = {std::get<1>(GetParam()), tested_data_type, "column_name", std::nullopt};
```
Comment (Boffmann) [inline src/test/synthetic_table_generator_test.cpp]: The last std::nullopt could be omitted here

File: src/benchmarklib/synthetic_table_generator.cpp [hunk 3]
```diff
@@ -50,28 +50,22 @@ namespace opossum {
 std::shared_ptr<Table> SyntheticTableGenerator::generate_table(const size_t num_columns, const size_t num_rows,
                                                                const ChunkOffset chunk_size,
                                                                const SegmentEncodingSpec segment_encoding_spec) {
-  auto table =
-      generate_table({num_columns, {ColumnDataDistribution::make_uniform_config(0.0, _max_different_value)}},
-                     {num_columns, {DataType::Int}}, num_rows, chunk_size,
-                     std::vector<SegmentEncodingSpec>(num_columns, segment_encoding_spec), std::nullopt, UseMvcc::No);
+  ColumnSpecification column_specification = {{ColumnDataDistribution::make_uniform_config(0.0, _max_different_value)},
```
Comment (Boffmann) [inline src/benchmarklib/synthetic_table_generator.cpp]: The last two std::nullopt could be omitted here

File: src/benchmarklib/synthetic_table_generator.cpp [hunk 4]
```diff
@@ -50,28 +50,22 @@ namespace opossum {
...
 }
 
 std::shared_ptr<Table> SyntheticTableGenerator::generate_table(
-    const std::vector<ColumnDataDistribution>& column_data_distributions,
-    const std::vector<DataType>& column_data_types, const size_t num_rows, const ChunkOffset chunk_size,
-    const std::optional<ChunkEncodingSpec>& segment_encoding_specs,
-    const std::optional<std::vector<std::string>>& column_names, const UseMvcc use_mvcc) {
+    const std::vector<ColumnSpecification>& column_specifications, const size_t num_rows, const ChunkOffset chunk_size,
+    const std::optional<ChunkEncodingSpec>& segment_encoding_specs, const UseMvcc use_mvcc) {
   Assert(chunk_size != 0ul, "cannot generate table with chunk size 0");
```
Comment (Bouncner) [inline src/benchmarklib/synthetic_table_generator.cpp]: Not your code, but please first letter in upper case and add a trailing dot.

Comment (birneamstiel) [reply src/benchmarklib/synthetic_table_generator.cpp]: Fixed in ba0d0c488f478054b6a1606a86dcd6e23d775dd5.

File: src/benchmarklib/synthetic_table_generator.cpp (hunk 3 above)
Comment (birneamstiel) [reply src/benchmarklib/synthetic_table_generator.cpp]: Resolved in 9425a461a6a5665302b3e1bf26d383b825b0ffc8.

File: src/test/synthetic_table_generator_test.cpp (hunk 2 above)
Comment (birneamstiel) [reply src/test/synthetic_table_generator_test.cpp]: Resolved in 9425a461a6a5665302b3e1bf26d383b825b0ffc8.

**Comment by birneamstiel** [comment]
@Boffmann @Bouncner FYI: We have worked on the things you commented on. Would be great if you could have another look and possibly approve.

File: src/benchmarklib/synthetic_table_generator.cpp [hunk 5]
```diff
@@ -149,6 +142,23 @@ std::shared_ptr<Table> SyntheticTableGenerator::generate_table(
...
           }
 
+          std::vector<bool> null_values;
+
+          /**
+           * If a ratio of to-be-created NULL values is given, fill the null_values vector used in the ValueSegment
+           * constructor in a regular interval based on the null_ratio with true.
+           */
+          if (column_specifications[column_index].null_ratio) {
+            null_values = std::vector<bool>(chunk_size, false);
```
Comment (Bouncner) [inline src/benchmarklib/synthetic_table_generator.cpp]: `resize()`?

File: src/benchmarklib/synthetic_table_generator.cpp [hunk 6]
```diff
@@ -149,6 +142,23 @@ std::shared_ptr<Table> SyntheticTableGenerator::generate_table(
...
+
+            const double step_size = 1.0 / column_specifications[column_index].null_ratio.value();
+            double current_row_offset = 0.0;
+            while (current_row_offset < chunk_size) {
+              null_values[static_cast<int>(std::round(current_row_offset))] = true;
```
Comment (Bouncner) [inline src/benchmarklib/synthetic_table_generator.cpp]: `static_cast<size_t>`?

File: src/benchmarklib/synthetic_table_generator.cpp [hunk 7]
```diff
@@ -164,14 +174,20 @@ std::shared_ptr<Table> SyntheticTableGenerator::generate_table(
...
+                std::make_shared<ValueSegment<ColumnDataType>>(create_typed_segment_values<ColumnDataType>(values));
+          }
 
           if (!segment_encoding_specs) {
             segments[column_index] = value_segment;
           } else {
-            segments[column_index] = ChunkEncoder::encode_segment(value_segment, column_data_types[column_index],
-                                                                  segment_encoding_specs->at(column_index));
+            segments[column_index] = ChunkEncoder::encode_segment(
+                value_segment, column_specifications[column_index].data_type, segment_encoding_specs->at(column_index));
```
Comment (Bouncner) [inline src/benchmarklib/synthetic_table_generator.cpp]: Any reason to exclude the encoding in the column specification?

File: src/benchmarklib/synthetic_table_generator.hpp [hunk 8]
```diff
@@ -58,17 +58,31 @@ struct ColumnDataDistribution {
...
 };
 
+struct ColumnSpecification {
+  ColumnSpecification(const ColumnDataDistribution& data_distribution, const DataType& data_type,
+                      const std::optional<std::string> name = std::nullopt,
+                      const std::optional<float> null_ratio = std::nullopt)
+      : data_distribution(data_distribution), data_type(data_type), name(name), null_ratio(null_ratio) {}
+
+  const ColumnDataDistribution data_distribution;
+
```
Comment (Bouncner) [inline src/benchmarklib/synthetic_table_generator.hpp]: Personally, I'd remove the empty lines between the declarations.

File: src/test/synthetic_table_generator_test.cpp [hunk 9]
```diff
@@ -24,12 +24,12 @@ TEST_F(SyntheticTableGeneratorTest, StringGeneration) {
 
 TEST_F(SyntheticTableGeneratorTest, ThrowOnParameterLengthMismatch) {
   auto table_generator = std::make_shared<SyntheticTableGenerator>();
-  const auto uniform_distribution = ColumnDataDistribution::make_uniform_config(0.0, 1.0);
+  auto uniform_distribution_0_1 = ColumnDataDistribution::make_uniform_config(0.0, 1.0);
```
Comment (Bouncner) [inline src/test/synthetic_table_generator_test.cpp]: Why did you remove the const?

File: src/test/synthetic_table_generator_test.cpp [hunk 10]
```diff
@@ -24,12 +24,12 @@ TEST_F(SyntheticTableGeneratorTest, StringGeneration) {
...
+  std::vector<ColumnSpecification> column_specs = {size_t{5}, {uniform_distribution_0_1, DataType::Double}};
+  std::vector<SegmentEncodingSpec> chunk_encoding_specs = {size_t{2}, {EncodingType::Dictionary}};
 
-  // vectors storing the column properties are expected to have the same length
-  ASSERT_THROW(table_generator->generate_table({uniform_distribution, uniform_distribution}, {DataType::Int}, 15, 10),
-               std::logic_error);
-  ASSERT_THROW(table_generator->generate_table({uniform_distribution}, {DataType::Int, DataType::Int}, 15, 10),
+  // vector storing the column properties ist expected to have the same length as vector storing the encoding
```
Comment (Bouncner) [inline src/test/synthetic_table_generator_test.cpp]: ist > is

File: src/test/synthetic_table_generator_test.cpp [hunk 11]
```diff
@@ -24,12 +24,12 @@ TEST_F(SyntheticTableGeneratorTest, StringGeneration) {
 
 TEST_F(SyntheticTableGeneratorTest, ThrowOnParameterLengthMismatch) {
   auto table_generator = std::make_shared<SyntheticTableGenerator>();
-  const auto uniform_distribution = ColumnDataDistribution::make_uniform_config(0.0, 1.0);
+  auto uniform_distribution_0_1 = ColumnDataDistribution::make_uniform_config(0.0, 1.0);
+  std::vector<ColumnSpecification> column_specs = {size_t{5}, {uniform_distribution_0_1, DataType::Double}};
```
Comment (Bouncner) [inline src/test/synthetic_table_generator_test.cpp]: I think you can use `column_specs{5, {uniform_distribution_0_1, DataType::Double}}` ... 
Besides, I am not really a fan of using curly brackets for constructor calls. This post summarizes my personal opinion quite well: https://stackoverflow.com/a/37228443/1147726. Not sure if there is a common accepted pattern when to use what.

File: src/benchmarklib/synthetic_table_generator.cpp (hunk 5 above)
Comment (birneamstiel) [reply src/benchmarklib/synthetic_table_generator.cpp]: Replaced in 36c31800

File: src/benchmarklib/synthetic_table_generator.hpp (hunk 8 above)
Comment (birneamstiel) [reply src/benchmarklib/synthetic_table_generator.hpp]: Done in 36c31800

File: src/test/synthetic_table_generator_test.cpp (hunk 9 above)
Comment (birneamstiel) [reply src/test/synthetic_table_generator_test.cpp]: Not on purpose, fixed in 36c31800

File: src/test/synthetic_table_generator_test.cpp (hunk 10 above)
Comment (birneamstiel) [reply src/test/synthetic_table_generator_test.cpp]: fixed in 36c31800

File: src/benchmarklib/synthetic_table_generator.cpp (hunk 6 above)
Comment (birneamstiel) [reply src/benchmarklib/synthetic_table_generator.cpp]: Replaced in 36c31800

File: src/test/synthetic_table_generator_test.cpp (hunk 11 above)
Comment (Bouncner) [reply src/test/synthetic_table_generator_test.cpp]: @mrks : we don't have a guideline for Hyrise here, right?

Comment (mrks) [reply src/test/synthetic_table_generator_test.cpp]: I was a fan of curly braces until I spent half an hour debugging a case where I used `vector{a, b}` and ended up with a vector of size 2. I still used curly braces (and [auto-to-stick](https://www.fluentcpp.com/2018/09/28/auto-stick-changing-style/), but avoid it for vector initializations). No hard rules here.

File: src/benchmarklib/synthetic_table_generator.cpp (hunk 7 above)
Comment (birneamstiel) [reply src/benchmarklib/synthetic_table_generator.cpp]: No, actually not... Included it in column specification in 710bf3f

File: src/test/synthetic_table_generator_test.cpp (hunk 11 above)
Comment (birneamstiel) [reply src/test/synthetic_table_generator_test.cpp]: So, what's the agreement here? This is the last open discussion I think...

Comment (Bouncner) [reply src/test/synthetic_table_generator_test.cpp]: As the post https://stackoverflow.com/a/37228443/1147726 recommends, I'd use `()` for constructor calls and `{}` for passing a data list. So I'd vote for `()` here, since it's calling a "complex" constructor and does not pass a data item of 5 here.

Comment (mrks) [reply src/test/synthetic_table_generator_test.cpp]: I'd do `auto column_specs = std::vector<ColumnSpecification>(...);`, but I don't think we should be blocked by this. @Bouncner, ready to go?

Comment (Bouncner) [reply src/test/synthetic_table_generator_test.cpp]: Yep, ready to go with that change.

Comment (mrks) [reply src/test/synthetic_table_generator_test.cpp]: Mmh, I was just about to change this myself so that we can merge this, but I don't have any rights on the PR. :/

Comment (birneamstiel) [reply src/test/synthetic_table_generator_test.cpp]: @mrks you have been granted the rights to our fork but you probably need to accept it here https://github.com/bakoe/hyrise/invitations

Comment (birneamstiel) [reply src/test/synthetic_table_generator_test.cpp]: RE this change: I just realized that including the encoding spec in our column specification makes this test case unnecessary so it's actually removed in https://github.com/hyrise/hyrise/pull/1987/commits/710bf3f25a8a0baa40682daa5d7426cb1de64b6e

Comment (mrks) [reply src/test/synthetic_table_generator_test.cpp]: Haha, ok
